### DISCUSSION TITLE
"Resolution"

### DISCUSSION THREAD

**Discussion by jkominek** [jkominek/piano-conversion#31] [General]
![time-space-resolution](https://user-images.githubusercontent.com/885262/104533550-d67c6580-55cf-11eb-8f10-04f0aeff16a6.png)

Here's some diagrams to help think about "resolution". Note the "Cybrid w/Interrupts" is hypothetical.

We've got time on the X axis, and position/distance on the Y axis. Using comparators or mechanical switches gets us coarse position resolution. How we read them (polling vs interrupts) gets us (relatively) slow or fast time resolution.

An ADC can get us finer position resolution, and it might be scanned quickly, or slowly, determining the time resolution.

But! We want hammer/key/MIDI **velocity**. Velocity corresponds to the slope of the line in the graphs. If we only have 3 position states ("far", "approaching", "close"), then our slope resolution is limited entirely by the time resolution. If we've got great position resolution and we check it only infrequently, then slope resolution is limited by the position.

Ideally, the design used by this project will have good time **and** space resolution, greatly increasing the the slope/velocity resolution.

The first firmware version will probably just use the ADC as a sort of "software-adjustable comparator". But we'll be able to refine that later, and look at not just when the "comparator set points" are crossed, but how far away from them the value was before/after the cross, to better establish velocity. (I'm not sure I've explained that clearly.)

I suspect the resolution won't be uniform across the space of velocities. Higher velocities will probably be coarser than low velocities. (Which I suspect is good, from a human perception stand point.) That is, many of the velocity code words corresponding to high velocities may never be generated, even if the code actually sends 32 bits of velocity.

That's worth elaborating on as well. MIDI 2.0 provides 16 bits for velocity. MIDI 1.0 with CC#88 High Resolution Velocity has 14 bits. And standard MIDI 1.0 has 7 bits.

I **will* send the maximum number of bits I can. If I implement MIDI 2.0, I **will** send 16 bit velocities. If I only do MIDI 1.0, I **will** send 14 bit velocities. But I can't guarantee (especially in all possible installations!) that it will be physically possible to press a key such that the code will generate all 16384 or 65536 velocities for any given key. There will probably be some gaps, and there will probably be more them at higher velocities.

**Comment by davidedelvento** [comment]
This explanation sounds very clear and exciting to me!

**Comment by ekumanov** [comment]
Very good comparison! My main concern while I was designing the Cybrid was exactly this problem. I didn't know how to use interrupts but I planned on switching to interrupts if the rough polling (as seen on the first picture) would turn out to be slow. I was worried though because I'm a high-level Java software engineer and all this low-level stuff with C-code is still rather enigmatic to me and I'm very inflexible in that regard. Switching to interrupts would mean I have to abandon the very simple Arduino-based approach and start reading some low-level specs and I didn't have time to learn an entirely new programming paradigm. Anyway, after I finished the first working prototype I made measurements with resolution calculations and for Teensy 3.6 it was about 1.5 MIDI velocity points at the highest velocities (or in other words a velocity of 125 can in fact be detected in a region of 123.5 - 126.5 but due to rounding to become 123 - 127, wow!). Sounds like not very precise, doesn't it? What amazed me though is that I couldn't perceive any lack of precision! Well, I may not be the most sensitive pianist in the world, who knows. But I think we're more sensitive around quieter dynamics and luckily the precision there was better. After I switched to Teensy 4.1 though, the precision at high velocities came down to 0.5 if I'm not mistaken, wheres as MIDI velocities of 64 it would be about 0.3 and that's already good enough for regular MIDI. With that in mind, I'm not a huge believer in MIDI 2.0 and the need for better precision, which is why I abandoned any plans on improving Cybrid further from this point.

However I appreciate the attention to detail and high resolution in this project, it's so professional I feel rather silly with the Cybrid 😀I guess when I have time some good day, I may try the Piano-Conversion project.

**Comment by davidedelvento** [comment]
@jkominek have you considered utilizing simultaneous sample ADCs such as https://www.microchip.com/en-us/product/MCP3914 (this is just one example of this class of devices, not a specific one, there are other chip from other manufacturers too)?

Benefits I see compared to the current approach:
- potentially better specs of the ADC compared to the multiplexed SAR
- allowing one to not be tied with one specific MCU which has *both* the computing and the ADC requirements
- consequence of the previous, one can select ADC and MCU separately, and there is a big number of products in stock (the ST32H43 seems the worst affected product by the shortage).
- freeing the dependency from ST32 which seems to have never recovered from the supply chain issues unlike other manufacturers such as Microchip (who has almost everything is stock and publishes number of items available today and date and number of next production runs -- however they don't have a great MCU with great ADC into a single package, but they do once one goes with two parts)

Problems I see with this approach:
- work needs to be redone for both hw and sw design
- overall increase in the complexity of the design, having more parts and communication to be involved
- the developer boards (equivalent of the Nucleo), at least from Microchip (haven't checked from TI) are much more sophisticated and hence much more expensive
- possibly other problems with the delta-sigma approach to ADC which might be deal breaker (but I don't know if that's the case)

I suspect you have considered this option already, and perhaps discarded it for reasons I ignore. If so, I was wondering if you'd want to share those reasons, and if they still apply in current context of ST part shortage (actually the parts are available and assembly houses would assemble them, but they would need to purchase the MCU from the sharks ending up with absurd prices, jacked up more than 20x the list price)

**Comment by davidedelvento** [comment]
See also https://pianoclack.com/forum/d/916-performance-of-diy-controller-prototype/9
